src/hardware/gus.cpp

```diff
@@ -1,924 +1,1418 @@
+{
+	uint8_t state = ctrl.state;
+	if (ctrl.irq_state & irq_mask)
+		state |= 0x80;
+	return state;
+}
+
+uint8_t Voice::ReadPanPot() const
+{
+	return pan_position;
+}
+
+uint8_t Voice::ReadVolState() const
+{
+	return ReadCtrlState(vol_ctrl);
+}
+
+uint8_t Voice::ReadWaveState() const
+{
+	return ReadCtrlState(wave_ctrl);
+}
+
+void Voice::ResetCtrls()
+{
+	vol_ctrl.pos = 0u;
+	UpdateVolState(0x1);
+	UpdateWaveState(0x1);
+	WritePanPot(PAN_DEFAULT_POSITION);
+}
+
+bool Voice::UpdateCtrlState(VoiceCtrl &ctrl, uint8_t state)
+{
+	const uint32_t orig_irq_state = ctrl.irq_state;
+	ctrl.state = state & 0x7f;
+	// Manually set the irq
+	if ((state & 0xa0) == 0xa0)
+		ctrl.irq_state |= irq_mask;
+	else
+		ctrl.irq_state &= ~irq_mask;
+
+	// Indicate if the IRQ state changed
+	return orig_irq_state != ctrl.irq_state;
+}
+
+bool Voice::UpdateVolState(uint8_t state)
+{
+	return UpdateCtrlState(vol_ctrl, state);
+}
+
+bool Voice::UpdateWaveState(uint8_t state)
+{
+	return UpdateCtrlState(wave_ctrl, state);
+}
+
+void Voice::WritePanPot(uint8_t pos)
+{
+	constexpr uint8_t max_pos = PAN_POSITIONS - 1;
+	pan_position = std::min(pos, max_pos);
+}
+
+// Four volume-index-rate "banks" are available that define the number of
+// volume indexes that will be incremented (or decremented, depending on the
+// volume_ctrl value) each step, for a given voice.  The banks are:
+//
+// - 0 to 63, which defines single index increments,
+// - 64 to 127 defines fractional index increments by 1/8th,
+// - 128 to 191 defines fractional index increments by 1/64ths, and
+// - 192 to 255 defines fractional index increments by 1/512ths.
+//
+// To ensure the smallest increment (1/512) effects an index change, we
+// normalize all the volume index variables (including this) by multiplying by
+// VOLUME_INC_SCALAR (or 512). Note that "index" qualifies all these variables
+// because they are merely indexes into the vol_scalars[] array. The actual
+// volume scalar value (a floating point fraction between 0.0 and 1.0) is never
+// actually operated on, and is simply looked up from the final index position
+// at the time of sample population.
+void Voice::WriteVolRate(uint16_t val)
+{
+	vol_ctrl.rate = val;
+	constexpr uint8_t bank_lengths = 63;
+	const int pos_in_bank = val & bank_lengths;
+	const int decimator = 1 << (3 * (val >> 6));
+	vol_ctrl.inc = ceil_sdivide(pos_in_bank * VOLUME_INC_SCALAR, decimator);
+
+	// Sanity check the bounds of the incrementer
+	assert(vol_ctrl.inc >= 0 && vol_ctrl.inc <= bank_lengths * VOLUME_INC_SCALAR);
+}
+
+void Voice::WriteWaveRate(uint16_t val)
+{
+	wave_ctrl.rate = val;
+	wave_ctrl.inc = ceil_udivide(val, 2u);
+}
+
+Gus::Gus(uint16_t port, uint8_t dma, uint8_t irq, const std::string &ultradir)
+        : port_base(port - 0x200),
+          dma2(dma),
+          irq1(irq),
+          irq2(irq)
+{
+	// Create the internal voice channels
+	for (uint8_t i = 0; i < MAX_VOICES; ++i) {
+		voices.at(i) = std::make_unique<Voice>(i, voice_irq);
```

Comment:
*Note for myself: investigate in here if `std:array` of `unique_ptrs` does not have some unfortunate implications…*